Keyboard End-to-end test helpers

1Copenut opened the pull request:
## Description
This is the much discussed but seldom seen keyboard-e2e test suite, part of it anyway. The focusable, tabbable, disabled element, and deep equality assertions use the [Ally.js library](https://allyjs.io/index.html).

I've included the basic helper functions for asserting [focusable](https://allyjs.io/api/query/focusable.html) and [tabbable](https://allyjs.io/api/query/tabbable.html) elements, and mocking keyboard navigation behavior, including:

* Asserting if a page or container has the same number of focusable and tabbable elements
* Asserting the count of focusable elements (useful on static pages)
* Asserting the count of tabbable elements (useful on static pages)
* Asserting if an element has a `disabled` attribute (useful for buttons)
* Mocking keyboard behavior for forward traversing radio buttons with `ARROW_DOWN` and `ARROW_RIGHT`
* Mocking keyboard behavior for reverse traversing radio buttons with `ARROW_UP` and `ARROW_LEFT`
* Mocking keyboard behavior for interacting with `<select>` menus
* ~~TODO:~~ Imported function for mocking keyboard behavior for interacting with checkboxes
* ~~TODO:~~ Created function for mocking keyboard behavior interacting with text inputs, text areas
* Created function for opening and closing modal windows
* Created a helper to avoid having to type `client.keys(client.Keys.TAB)` repeatedly
* TODO: Add documentation page to highlight good cases for these tests

These are different from existing e2e test helpers. The functions make extensive use of `client.keys(KEY)` to simulate `TAB`, `ENTER`, and `ARROW` key behavior, just like a real user navigating with their keyboard. These tests should be written as support for happy path end-to-end tests.

## Testing done
* Full test suite run locally
* Keyboard-e2e test put together for the GI Bill Comparison Tool
* Keyboard-e2e test put together for the Education Feedback Tool

## Acceptance criteria
- [x] Tests pass for keyboard-e2e helpers

## Definition of done
- [ ] Events are logged appropriately
- [x] Documentation has been updated, if applicable
- [x] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)
- [x] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

Commit:
First commit, adding helper methods and proof test.

Commit:
Adding methods for radio buttons, checking disabled elements.

Commit:
Confirming the allyDeepEquals assertion works

Commit:
Removing a few static checks because of the React app dynamic load.

Commit:
Removing a wayward class.

1Copenut commented:
/cc @pjhill && @tlei123 for awareness. The keyboard-e2e suite is now under draft PR, with a few items left to build. These test helpers are blueprinted in the GI Bill Comparison Tool test included in the PR.

Commit:
Removing an extra step from the select helper.

1Copenut commented on src/platform/testing/e2e/nightwatch-commands/evaluateSelectMenu.js:
```diff
@@ -0,0 +1,37 @@
+) {
+  const client = this;
+
+  return client.waitForElementVisible(selectMenu, 1000, () => {
```
Change this to `waitForElementPresent`

1Copenut commented on src/platform/testing/e2e/nightwatch-commands/allyDeepEquals.js:
```diff
@@ -0,0 +1,45 @@
+ */
+exports.command = function allyDeepEquals(selector, callback) {
+  return this.execute(
+    function(sel) {
```
Refactor this to an `() =>` arrow function

1Copenut commented on src/platform/testing/e2e/nightwatch-commands/allyCheckTabbableCount.js:
```diff
@@ -0,0 +1,33 @@
+ */
+exports.command = function allyCheckTabbableCount(selector, callback) {
+  return this.execute(
+    function(sel) {
```
Refactor this to an `() =>` arrow function

cvalarida replied:
You may or may not run into trouble with this. Specifically, [arrow functions](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions) change the normal rebinding of `this`. Maybe it'll be fine, but if you go to refactor it and run into unexpected issues, that might be why.

Also just throwing it out there: Just because it _can_ be changed to be "sleeker" doesn't mean it necessarily _should_ be. Having `function` declarations is totally fine, so don't feel like you need to in order for this to be "clean."

1Copenut replied:
That's true. I think it was more about being consistent. I had one arrow function in the group and it worked okay, but yeah, moving back to `function(sel)` doesn't rebind `this`, and that may be more beneficial.

1Copenut commented on src/platform/testing/e2e/nightwatch-commands/allyCheckFocusableCount.js:
```diff
@@ -0,0 +1,35 @@
+ */
+exports.command = function allyCheckFocusableCount(selector, callback) {
+  return this.execute(
+    function(sel) {
```
Refactor this to an `() =>` arrow function

Commit:
WIP. Refactoring and adding helper methods.

cvalarida commented on src/platform/testing/e2e/nightwatch-commands/evaluateRadioButtons.js:
```diff
@@ -0,0 +1,35 @@
+  return client.waitForElementPresent(element, 1000, function() {
+    for (let i = 0; i < arrLength; i += 1) {
+      this.assert.isActiveElement(selectorArray[i]).keys(arrowPressed);
+    }
```
Alternatively:
```js
selectorArray.forEach(s => this.assert.isActiveElement(s).keys(arrowPressed));
```

cvalarida commented on src/platform/testing/e2e/nightwatch-commands/evaluateRadioButtonsReverse.js:
```diff
@@ -0,0 +1,38 @@
+  return client.waitForElementPresent(element, 1000, function() {
+    for (let i = arrLength - 1; i >= 0; i -= 1) {
+      this.keys(arrowPressed).assert.isActiveElement(selectorArray[i]);
+    }
```
Alternatively:
```js
selectorArray
  .reverse()
  .forEach(s => this.keys(arrowPressed).assert.isActiveElement(s));
```

cvalarida replied:
Ideally, I think you wouldn't care about the order, but I don't understand what this is doing, so I assume you have to care. :man_shrugging:

1Copenut replied:
Radio buttons are sequential when you're interacting with them by keypress. These methods are gathering the array, asserting each index is focused, pressing the arrow key, then re-asserting focus. I hadn't seen the `reverse()` method before, so having that is nice and allows for more functional means of achieving the same goal.

cvalarida commented:
I have a lot of thoughts on this!

Before I vomit those up, though, how important would you say it is to have these kinds of keyboard navigation tests? What kinds of questions are you trying to answer with these?

**Context:** I think we can abstract this out a bit further to reduce the burden on the folks building the tests, but before I can give any kind of good suggestions in that regard, I need a better understanding of the underlying reasons.

1Copenut commented:
@cvalarida I appreciate the look at these. I've got a lot of open time on my calendar tomorrow if you'd like to schedule some to discuss intent and ways we could abstract further. Thank you.

Commit:
Merge branch 'master' into feature/keyboard-e2e-proof

Commit:
Added helpers for inputs, modals, checkboxes, repeated keys.

Commit:
Refactoring away from arrow functions for a few helpers.

1Copenut commented:
@cvalarida I'm ready for this to be reviewed for a 1.0 beta release. I've added a few helpers, and another test for the Education feedback tool. That one was useful for checkboxes, and needing to handle multiple keypresses for `TAB` to move through a swath of checkboxes quickly. I'm looking forward to piloting this with a team immediately, and getting to work on the second iteration that abstracts at a level higher per our discussion last week.

Commit:
Updating docs for repeatKeypress helper.

Commit:
Moving the keyboard test for feedback tool to use slower timeouts.

Commit:
Revising a selector on education feedback form.

Commit:
Refactoring to tab and submit forms, not shortcut with ENTER.

1Copenut commented:
I'm piloting this with the Facility Locator team starting tomorrow and would like to have this code merged if possible before that discussion. If not, will need to get it merged in early next week and I can speak to it using local feature branches.

cvalarida commented on src/platform/testing/e2e/nightwatch-assertions/isAllyDeepEquals.js:
```diff
@@ -0,0 +1,31 @@
+ * @param {boolean} [balanceBool] True if page has no no tabIndex="-1"
+ * @api assertions
+ */
+exports.assertion = function isAllyDeepEquals(selector, balanceBool) {
```
:thinking: Can you help me understand what this is supposed to be doing. I'm not following its intended purpose, I think.

Looking at the related command (`allyDeepEquals`), it looks like this asserts that the number of tabbable elements matches the number of focusable elements within the `selector`. Is that right?

It's also worth noting that even if there _are_ the same number of each, that doesn't mean they're the same _set_ of elements. Consider:
```
Tabbable elements:
[ a b c ]

Focusable elements:
[ a b d ]
```

1Copenut replied:
@cvalarida I'm going to remove this function. Its usefulness was less and less clear as I started to work on this feedback. I've updated the focusable and tabbable functions to remove the external library and be more expressive.

cvalarida commented on src/platform/testing/e2e/nightwatch-assertions/isAllyFocusableCount.js:
```diff
@@ -0,0 +1,26 @@
+ * @param {string} [message] Optional log message displayed in output. Defaults to this.message.
+ * @api assertions
+ */
+exports.assertion = function isAllyFocusableCount(selector, count, message) {
```
What do you think of calling this something like `hasFocusableCount`? It's really asserting that it _has_ a certain number of focusable elements, not that it _is_ a focusable count, y'know?

cvalarida commented on src/platform/testing/e2e/nightwatch-assertions/isAllyTabbableCount.js:
```diff
@@ -0,0 +1,27 @@
+ * @param {string} [message] Optional log message displayed in output. Defaults to this.message.
+ * @api assertions
+ */
+exports.assertion = function isAllyTabbableCount(selector, count, msg) {
```
What do you think about calling this `hasTabbableCount`?

cvalarida commented on src/platform/testing/e2e/nightwatch-commands/allyCheckFocusableCount.js:
```diff
@@ -0,0 +1,35 @@
+ * @returns {array} The array of focusable elements found in the selector.
+ * @api commands
+ */
+exports.command = function allyCheckFocusableCount(selector, callback) {
```
Is this command useful outside the context of the assertion? If not, consider moving inside the assertion function.

cvalarida commented on src/platform/testing/e2e/nightwatch-commands/allyCheckTabbableCount.js:
```diff
@@ -0,0 +1,33 @@
+ * @returns {array} The array of tabbable elements found in the selector.
+ * @api commands
+ */
+exports.command = function allyCheckTabbableCount(selector, callback) {
```
Is this command useful outside the context of the assertion? If not, consider moving inside the assertion function.

cvalarida commented on src/platform/testing/e2e/nightwatch-commands/allyDeepEquals.js:
```diff
@@ -0,0 +1,45 @@
+ * @returns {array} [comparator] The array objects for focusable, and tabbable elements.
+ * @api commands
+ */
+exports.command = function allyDeepEquals(selector, callback) {
```
Is this command useful outside the context of the assertion? If not, consider moving inside the assertion function.

cvalarida commented on src/platform/testing/e2e/nightwatch-commands/checkDisabledElement.js:
```diff
@@ -0,0 +1,23 @@
+ * @returns {boolean}
+ * @api commands
+ */
+exports.command = function checkDisabledElement(selector, callback) {
```
Is this command useful outside the context of the assertion? If not, consider moving inside the assertion function.

1Copenut replied:
Works much better this way!

cvalarida commented on src/platform/testing/e2e/nightwatch-commands/evaluateCheckboxes.js:
```diff
@@ -0,0 +1,41 @@
+  const client = this;
+
+  return client.waitForElementPresent(element, 1000, function() {
+    for (let i = 0; i < arrLength; i += 1) {
```
Alternatively:
```js
arrLength.forEach((element, index) => {
  ...
```

cvalarida commented on src/platform/testing/e2e/nightwatch-commands/evaluateInput.js:
```diff
@@ -0,0 +1,29 @@
+ * @param {string} inputText The text string that should be keyed into the input.
+ * @api commands
+ */
+module.exports.command = function evaluateInput(input, inputText) {
```
This looks like it more or less duplicates the `fill` custom command. Would it be desirable to add the active element and end value checks to that command instead?

cvalarida commented on src/platform/testing/e2e/nightwatch-commands/evaluateModalWindow.js:
```diff
@@ -0,0 +1,42 @@
+ * @param {object} [triggerKey] The client.Keys.KEY being pressed to open and close the modal window. Default key is ENTER.
+ * @api commands
+ */
+module.exports.command = function evaluateModalWindow(
```
What do you think about `allyEvaluateModalWindow` to make it clearer _what_ it's evaluating?

cvalarida commented on src/platform/testing/e2e/nightwatch-commands/evaluateRadioButtons.js:
```diff
@@ -0,0 +1,35 @@
+ * @param {object} arrowPressed Nightwatch Keys object. Expects ARROW_DOWN || ARROW_RIGHT.
+ * @api commands
+ */
+exports.command = function evaluateRadioButtons(selectorArray, arrowPressed) {
```
What do you think about `allyEvaluateRadioButtons` to make it clearer what it's evaluating?

cvalarida commented on src/platform/testing/e2e/nightwatch-commands/evaluateRadioButtonsReverse.js:
```diff
@@ -0,0 +1,38 @@
+ * @param {object} [arrowPressed] Nightwatch Keys object. Expects ARROW_UP || ARROW_LEFT.
+ * @api commands
+ */
+exports.command = function evaluateRadioButtonsReverse(
```
Given that we have a `.reverse()`, what do you think about adding an option to the `evaluateRadioButtons` command that reverses the order?

cvalarida commented on src/platform/testing/e2e/nightwatch-commands/evaluateSelectMenu.js:
```diff
@@ -0,0 +1,37 @@
+ * @param {string} selectedOption Value attribute of the <option> that should be selected.
+ * @api commands
+ */
+module.exports.command = function evaluateSelectMenu(
```
Same story as the others. What do you think about `allyEvaluateSelectMenu`?

cvalarida commented on src/platform/testing/e2e/nightwatch-commands/repeatKeypress.js:
```diff
@@ -0,0 +1,28 @@
+ * @param {number} pressCount Number of times to press the key.
+ * @api commands
+ */
+module.exports.command = function repeatKeypress(key, pressCount) {
```
I'd recommend adding another optional argument for `waitTime` and give it a default of 125 rather than just hardcoding that bit in there.

cvalarida reviewed (commented):
:wave: Gave it a more thorough review this go-round!

Commit:
WIP. Refactoring the focusable and tabbable assertions.

Commit:
Renaming keyboard evaluation functions.

Commit:
Refactoring the evaluateRadioButtons into one function.

Commit:
WIP. Refactoring focusable, tabbable assertions and docs.

Commit:
Refactoring education-feedback test and repeat keys helper.

Commit:
Removing ally.js from lockfile.

1Copenut commented:
@cvalarida && @brangi This is ready for one more time around review. I've refactored somewhat critically and cut down over 100 lines, and 6 files. Also removed a dependency by rolling my own more specific `focusable` and `tabbable` assertions.

Commit:
Refactoring hasFocusable and hasTabbable for clarity.

Commit:
Merge branch 'master' into feature/keyboard-e2e-proof

1Copenut commented:
@cvalarida && @brangi If either of you have a few minutes, would you mind reviewing one more time? I'd like to get these merged into `master` so we can pilot and iterate on more advanced keyboard tests. Thanks y'all!

/cc @bkjohnson

bkjohnson commented on src/applications/gi/tests/01-keyboard.e2e.spec.js:
```diff
@@ -0,0 +1,153 @@
+  // This one is a bit unique because the second radio is pre-checked.
+  client
+    .keys(TAB)
+    .keys(TAB)
```
Another opportunity for `repeatKeypress`

bkjohnson commented on src/applications/gi/tests/01-keyboard.e2e.spec.js:
```diff
@@ -0,0 +1,153 @@
+  // Let's try to submit an incomplete form by skipping over the city typeahead
+  client
+    .keys(TAB)
+    .keys(TAB)
```
_cue DJ Khaled_ 
> And another one

bkjohnson commented on src/applications/edu-benefits/tests/feedback-tool/01.feedback-tool-keyboard.e2e.spec.js:
```diff
@@ -0,0 +1,224 @@
+        'input#root_onBehalfOf_0',
+        'input#root_onBehalfOf_1',
+        'input#root_onBehalfOf_2',
+      ],
```
I'm a bit torn.  On the one hand, I see the value in being specific with each selector so it is clear exactly what the helper will look at.  However, I'm also wondering if it would be a better experience to just specify the first radio button in a group (or specify the container) and have the helper take care of finding the sibling inputs, like this: 

```js
    .allyEvaluateRadioButtons(
      'div#button-group',
      ARROW_UP,
      true,
    );
```

 I say this with no idea of what the DOM looks like though, so it may not be a good idea.

1Copenut replied:
@bkjohnson This gives me an idea: https://developer.mozilla.org/en-US/docs/Web/API/Element/getElementsByTagName :)

bkjohnson reviewed (commented):
A few comments and questions.  Also, it was enlightening to see the subtle differences between hasTabbleCount` and `hasFocusableCount`.

Commit:
WIP. Refactoring time to wait for selector.

1Copenut commented on src/platform/testing/e2e/nightwatch-commands/allyEvaluateCheckboxes.js:
```diff
@@ -0,0 +1,40 @@
+      this.assert
+        .isActiveElement(sel)
+        .keys(SPACE)
+        .assert.attributeContains(element, 'checked', 'true');
```
*Excellent* point sir. That was indeed a bug in the system. Refactoring this afternoon.

Commit:
WIP. Refactoring a bug in checkbox helper logic.

1Copenut commented on src/platform/testing/e2e/nightwatch-commands/allyEvaluateRadioButtons.js:
```diff
@@ -0,0 +1,48 @@
+        this.assert.isActiveElement(sel).keys(arrowPressed),
+      );
+    }
+  });
```
I like the separate logic, because the array has to be reversed when you're checking radio buttons for `UP_ARROW` and `LEFT_ARROW` handling. This is more verbose, but is also more explicit.

I checked the test output and it's asserting that all radios are focused on the reversed order.

Commit:
Removing top-level arrow functions, adding argument defaults.

1Copenut commented on src/platform/testing/e2e/nightwatch-commands/allyEvaluateSelectMenu.js:
```diff
@@ -0,0 +1,37 @@
+) {
+  const client = this;
+
+  return client.waitForElementPresent(selectMenu, 1000, () => {
```
@bkjohnson I like this suggestion. I bumped the default to 2000 ms and added an optional argument to pass a different amount of time.

1Copenut commented on src/platform/testing/e2e/nightwatch-commands/allyEvaluateModalWindow.js:
```diff
@@ -0,0 +1,42 @@
+/* eslint-disable func-names, prefer-arrow-callback */
+
+/**
+ * Evaluate modal windows for basic keyboard open and close functionality:
```
I think there's situations where that will be good to have, and plan to build for that in the near term. I'd like to leave this as a basic "smoke test" r/n and get the first iteration published so teams can test with it and extend as they find a need.

Clark-W-Griswold committed:
get keyboard e2e test working for gi app

Clark-W-Griswold committed:
Merge branch 'master' of github.com:department-of-veterans-affairs/vets-website into feature/keyboard-e2e-proof

Clark-W-Griswold committed:
Merge branch 'master' of github.com:department-of-veterans-affairs/vets-website into feature/keyboard-e2e-proof

Clark-W-Griswold committed:
update selectors to work with changing id's

Clark-W-Griswold committed:
lint fix

Clark-W-Griswold committed:
Merge branch 'master' of github.com:department-of-veterans-affairs/vets-website into feature/keyboard-e2e-proof

cvalarida commented on src/platform/testing/e2e/nightwatch-commands/allyEvaluateInput.js:
```diff
@@ -0,0 +1,29 @@
+/* eslint-disable func-names, prefer-arrow-callback */
```
I think these are fine to disable.
- `func-names`: For stack traces, it's better for the function to be called `allyEvaluateInput` than `command`
- `prefer-arrow-callback`: These Nightwatch custom commands and assertions use `this`, which has a different scope in an arrow function

cvalarida reviewed (approved):
I'd still like to see the custom Nightwatch commands only used in the custom assertions moved to inside the assertion functions.

1Copenut commented:
> I'd still like to see the custom Nightwatch commands only used in the custom assertions moved to inside the assertion functions.

Agreed. I think the only one this still applies to is the `isDisabled` custom assertion.

Clark-W-Griswold committed:
adding newline to trigger va-vfs-bot in github?

shawkey-daniel reviewed (approved):
LGTM
